### PR TITLE
Dylan/onboarding-upgrades

### PR DESCRIPTION
**Pull request by dylanh724**
## Problem

Onboarding towards the path of C# and Unity experiences a minor block, alongside minor inconsistencies, bloat and room for UX improvements.

## Solution

- Fixed a blocking syntax error with the "Hello World" CLI example.
- Fixed broken .unitypackage URL in C# ref.
- Polished with minor typo fixes, code consistency (basing on other examples), more internal reference links.
- Additional warning regarding C# speed vs Rust.
- In the Unity tutorial, we go the path of Rust for the module: Unity devs are more-likely wanting C# for a quickstart (since we 100% know they know C#, but unsure regarding Rust):
  - Linked to C# Module tutorial before we jump to Rust, as a placeholder.
  - For pt2 (a future PR), I'll work with @jdetter to parity the Rust code in C# to diverge tutorial a bit (or perhaps have code tabs, as John suggested).
- Renamed `Unity3d` refs to the modern-named `Unity` and added "or Rider" appended to _Visual Studio_ mentions.
- Added `OneTimeQuery` C# api ref.
- Related, I [opened another PR](https://github.com/clockworklabs/spacetime-web/pull/358) on `spacetime-web` that fixes all external links that are currently showing as unstyled.
- [Verbose Details with Screenshots (Google Doc - Clockwork Only)](https://docs.google.com/document/d/1qwzb9yZ16NQ_KoO_QdEE9KiKHR43EkcXQWedzMfWX_A/edit?usp=sharing)

## Testing

There was a README note instructing how to test below, but there seemed to be no instructions below (I also removed this note). However, I was able to use the preview pane in Rider to ensure expected visuals.

(**Edit:** Squashed for clarity)

### CODE REVIEW

File: docs/unity/part-1.md [hunk 1]
```diff
@@ -105,6 +112,8 @@ At this point you should have the single player game working. In your CLI, your
 spacetime start
 ```
 
+💡Below examples Rust language, [but you may also use C#](../modules/c-sharp/index.md) at the cost of 1/2 the speed.
```
Comment (bfops) [inline docs/unity/part-1.md:115]: I'm not sure that we should advertise/codify this speed difference :thinking: it's kind of in flux. @jdetter do you have thoughts?

File: docs/unity/part-1.md [hunk 2]
```diff
@@ -31,9 +36,9 @@ Open Unity and create a new project by selecting "New" from the Unity Hub or goi
 
 ![UnityHub-NewProject](/images/unity-tutorial/UnityHub-NewProject.JPG)
 
-For Project Name use `client`. For Project Location make sure that you use your `SpacetimeDBUnityTutorial` directory. This is the directory that we created in a previous step.
+**⚠️ Ensure `3D (URP)` is selected** to properly render the materials in the scene!
```
Comment (bfops) [inline docs/unity/part-1.md:39]: can we keep the word **Important** in here?

also, is this related to the shader issue you mentioned, or is it separate? you said that wasn't URP related right?

File: docs/unity/part-1.md [hunk 3]
```diff
@@ -12,14 +12,19 @@ This tutorial has been tested against UnityEngine version 2022.3.4f1. This tutor
 
 ## Prepare Project Structure
 
-This project is separated into two sub-projects, one for the server (module) code and one for the client code. First we'll create the main directory, this directory name doesn't matter but we'll give you an example:
+This project is separated into two sub-projects;
+
+1. Server (module) code 
+2. Client code.
```
Comment (bfops) [inline docs/unity/part-1.md:18]: trailing `.`?

File: docs/unity/part-1.md [hunk 4]
```diff
@@ -12,14 +12,19 @@ This tutorial has been tested against UnityEngine version 2022.3.4f1. This tutor
...
+
+First, we'll create an arbitrarily-named project root directory:
```
Comment (bfops) [inline docs/unity/part-1.md:20]: What do you think about this?
> First, we'll create a project root directory (you can choose the name):

For some reason, the word "arbitrary" in this context makes me think that the naming shouldn't matter to _me_, either.

File: docs/unity/part-1.md [hunk 5]
```diff
@@ -639,21 +649,26 @@ using SpacetimeDB;
...
-    Reducer.UpdatePlayerPosition(new StdbVector3
-        {
-            X = p.x,
-            Y = p.y,
-            Z = p.z,
-        },
-        PlayerMovementController.Local.GetModelRotation(),
-        PlayerMovementController.Local.IsMoving());
+   bool hasUpdatedRecently = lastUpdateTime.HasValue && 
+   Time.time - lastUpdateTime.Value > 1.0f / movementUpdateSpeed;
```
Comment (bfops) [inline docs/unity/part-1.md:653]: I like breaking up these variables :100: 

..but this line break/indent is super confusing :sweat_smile:

Comment (bfops) [reply docs/unity/part-1.md:653]: also, you didn't introduce this issue, but since we're here: this code is wrong!

Maybe we can add another variable?

```rust
   
   float? deltaTime = Time.time - lastUpdateTime;
   bool hasUpdatedRecently = deltaTime.HasValue && deltaTime.Value < 1.0f / movementUpdateSpeed;
   bool isConnected = SpacetimeDBClient.instance.IsConnected();

   if (hasUpdatedRecently || !isConnected)
   {
      return;
   }
```

File: docs/modules/c-sharp/quickstart.md [hunk 6]
```diff
@@ -250,12 +265,16 @@ From the `quickstart-chat` directory, run:
...
+npm i wasm-opt -g
+```
+
 ## Call Reducers
 
 You can use the CLI (command line interface) to run reducers. The arguments to the reducer are passed in JSON format.
 
 ```bash
-spacetime call <module-name> send_message '["Hello, World!"]'
+spacetime call <module-name> send_message "Hello, World!"
```
Comment (bfops) [inline docs/modules/c-sharp/quickstart.md:277]: I assume it caused some cryptic error with the square brackets? :sweat_smile:

**Review by bfops — CHANGES_REQUESTED** [review]
Nice, these are largely good improvements :+1: 

Left a few specific questions/concerns below

File: docs/unity/part-1.md [hunk 7]
```diff
@@ -639,21 +649,26 @@ using SpacetimeDB;
...
+   }
+
+   lastUpdateTime = Time.time;
+   var p = PlayerMovementController.Local.GetModelPosition();
+   
+   Reducer.UpdatePlayerPosition(new StdbVector3
+   {
+      Y = p.y,
+      Z = p.z,
+      X = p.x,
```
[the remark this answers is not in the document]
Comment (dylanh724) [reply docs/unity/part-1.md:668]: > why did we move this?

Must've been an inadvertent hotkey during the respacing; moving back! This block had +1 more horizontal space more than all the others.

File: docs/sdks/c-sharp/quickstart.md [hunk 8]
```diff
@@ -39,7 +39,7 @@ mkdir -p client/module_bindings
 spacetime generate --lang csharp --out-dir client/module_bindings --project-path server
 ```
 
-Take a look inside `client/module_bindings`. The CLI should have generated five files:
+Take a look inside `client/module_bindings`. The CLI should have generated five files under `module_bindings`:
```
[the remark this answers is not in the document]
Comment (dylanh724) [reply docs/sdks/c-sharp/quickstart.md:42]: > idk, I feel like this one was clear before and ends up kinda repetitive now?

Hmm, you're right ☕ removing

File: docs/unity/part-1.md (hunk 1 above)
Comment (dylanh724) [reply docs/unity/part-1.md:115]: There's a chance it could sound like we're passively-aggressively pushing them towards Rust, but at the same time... this would probably explain why a Unity tutorial shows Rust server code instead of C#?

File: docs/unity/part-1.md (hunk 2 above)
Comment (dylanh724) [reply docs/unity/part-1.md:39]: Sure, adding back **important** - unrelated (URP issues show all magenta)

File: docs/unity/part-1.md (hunk 4 above)
Comment (dylanh724) [reply docs/unity/part-1.md:20]: Takes focus off the name - better (changing to this)

File: docs/modules/c-sharp/quickstart.md (hunk 6 above)
Comment (dylanh724) [reply docs/modules/c-sharp/quickstart.md:277]: 'twas a cryptic err ;)

File: docs/unity/part-1.md (hunk 5 above)
Comment (dylanh724) [reply docs/unity/part-1.md:653]: Got it -- also, I'm wondering how you folks feel about this:

```csharp
 if (hasUpdatedRecently || !isConnected)
      return;
```

Do you like the shortcut versions for simple `if`s? I didn't touch these, just in case.

**Comment by dylanh724** [comment]
✅ Went through all the raised issues

**Comment by dylanh724** [comment]
image

New commit: Filter and Find being shown interchangeably in C# tutorial. Normalized to Find instead of Filter, since discussing just looking if 1 user exists.

**Comment by dylanh724** [comment]
Edited in the Google Doc draft with screenshots and verbose details:
https://docs.google.com/document/d/1qwzb9yZ16NQ_KoO_QdEE9KiKHR43EkcXQWedzMfWX_A/edit

**Comment by dylanh724** [comment]
Additionally pushed copy+paste redundant comments in Unity pt1:
image

**Comment by dylanh724** [comment]
**New minor commit:** Removed a Unity-pt1 `using` directive, and merged some info with another using statement that seems to now include the one removed:
image

**Comment by dylanh724** [comment]
**New minor commit:** Better code spacing for a crunched block
image

File: docs/getting-started.md [hunk 9]
```diff
@@ -3,14 +3,15 @@
...
-2. Run the start command
+2. Run the start command:
 
 ```bash
 spacetime start
 ```
 
 The server listens on port `3000` by default, customized via `--listen-addr`.
 
+💡 Standalone mode will run in the foreground: new commands will require a new terminal window.
```
Comment (PuruVJ) [inline docs/getting-started.md:14]: How about writing in first person: `Standalone mode will run in the foreground. Open a new terminal to run new commands`

It feels easier to parse

Comment (dylanh724) [reply docs/getting-started.md]: That's a good one, too. In the end, I changed it to "💡 Standalone mode will run in the foreground", since there are numerous ways to do this (someone pointed out you can also add `&` to make it run in the bg). Perhaps we don't need too much info, as long as they know the gist?

**Comment by dylanh724** [comment]
**New minor commit:** Removed speed mention, added foreground memo for standalone mode in Unity pt1 (since we're going to issue 2 commands one-after-another).
image

File: docs/unity/part-1.md (hunk 1 above)
Comment (dylanh724) [reply docs/unity/part-1.md]: For now, I removed the speed memo, but left the C# link

**Comment by dylanh724** [comment]
There's another fix I want to add, but didn't yet since it's a bit more complex:

image

* Swapping `PlayerComponent` to nullable with `FindBy` instead of `FilterBy` will fix a syntax err. This part is straight-forward.
* Then, we probably want to check if `is not null`, but this suddenly makes me realize: What if it's null?
  * In the tutorial, we never *actually used* the CreatePlayer Reducer, although we did create it. Is this the place to use if the PlayerComp is null? CC @bfops

**Comment by dylanh724** [comment]
> There's another fix I want to add, but didn't yet since it's a bit more complex: image
> 
> * Swapping `PlayerComponent` to nullable with `FindBy` instead of `FilterBy` will fix a syntax err. This part is straight-forward.
> * Then, we probably want to check if `is not null`, but this suddenly makes me realize: What if it _is_ null?
>   
>   * In the tutorial, we never _actually used_ the CreatePlayer Reducer, although we did create it. Is this the place to use if the PlayerComp is null? CC @bfops
> 
> ## Edit: How about this?
> image

Discussion: https://discord.com/channels/931210784011321394/1205409581120102451

File: docs/unity/part-1.md [hunk 10]
```diff
@@ -639,21 +650,26 @@ using SpacetimeDB;
...
-    Reducer.UpdatePlayerPosition(new StdbVector3
-        {
-            X = p.x,
-            Y = p.y,
-            Z = p.z,
-        },
-        PlayerMovementController.Local.GetModelRotation(),
-        PlayerMovementController.Local.IsMoving());
+   float? deltaTime = Time.time - lastUpdateTime;
+   bool hasUpdatedRecently = deltaTime.HasValue && deltaTime.Value < 1.0f / movementUpdateSpeed;
```
Comment (bfops) [inline docs/unity/part-1.md:654]: (review note: I asked to switch this to `<`, because as far as I can tell the `>` was a bug)

File: docs/unity/part-1.md (hunk 5 above)
Comment (bfops) [reply docs/unity/part-1.md]: I'm indifferent - @jdetter does it matter to you?

**Comment by dylanh724** [comment]
Final edit, following the resolution of the **FilterBy vs FindBy** discussion:
image

Just need review now :)

**Comment by dylanh724** [comment]
Just a quick bump ping to @bfops / @jdetter for review when free. No rush (I'm working on other projects while waiting).

**Review by bfops — APPROVED** [review]
Oops, thanks for the bump - meant to approve this! These improvements look great :+1: 

But let's **not merge** until John's had a chance to glance.

File: docs/unity/part-1.md [hunk 11]
```diff
@@ -553,7 +561,17 @@ public class RemotePlayer : MonoBehaviour
...
+        PlayerComponent? playerComp = PlayerComponent.FilterByEntityId(EntityId).FirstOrDefault();
+        if (playerComp is null)
+        {
+            string inputUsername = UsernameElement.Text;
+            Debug.Log($"PlayerComponent not found - Creating a new player ({inputUsername})");
+            Reducer.CreatePlayer(inputUsername);
+            
+            // Try again, optimistically assuming success for simplicity
+            PlayerComponent? playerComp = PlayerComponent.FilterByEntityId(EntityId).FirstOrDefault();
+        }
```
Comment (bfops) [inline docs/unity/part-1.md:573]: was this change necessary for some reason? It seems like a bug to have somehow created a `RemotePlayer` with an invalid `EntityId`?

**Review by bfops — COMMENTED** [review]
I think this looks good. We synced with @jdetter in a meeting as well.

One last question

File: docs/unity/part-1.md (hunk 11 above)
Comment (dylanh724) [reply docs/unity/part-1.md:573]: > was this change necessary for some reason? It seems like a bug to have somehow created a `RemotePlayer` with an invalid `EntityId`?

You're right! This change was unnecessarily, forgetting this was a _remote_ player. Upon this section revert, it may be worth renaming `playerComp` to `remotePlayerComp`? if I can miss it, others may too.

Comment (dylanh724) [reply docs/unity/part-1.md]: image

**Comment by dylanh724** [comment]
After being informed there was only 1 remaining question (answered+fixed), **merging!**
